### DISCUSSION TITLE
TypeError: The values of index must be integers

### DISCUSSION THREAD

**Discussion by hopegithub** [RapidAI/TableStructureRec#2] [Q&A] [answered]
我运行了作者提供的examle中无表格线示例，但报错，我查看我的输入没有问题，是一张表格图片
代码：
```python
# -*- encoding: utf-8 -*-
# @Author: SWHL
# @Contact: liekkaskono@163.com
from pathlib import Path

from lineless_table_rec import LinelessTableRecognition

engine = LinelessTableRecognition()

img_path = "../test11.jpg"
table_str, elapse = engine(img_path)

print(table_str)
print(elapse)

with open(f"{Path(img_path).stem}.html", "w", encoding="utf-8") as f:
    f.write(table_str)

print("ok")
```

报错：
```bash
WARNING:root:Traceback (most recent call last):
  File "D:\anaconda\envs\paddleocr\lib\site-packages\lineless_table_rec\main.py", line 54, in __call__
    polygons, slct_logi = self.infer(input_info)
  File "D:\anaconda\envs\paddleocr\lib\site-packages\lineless_table_rec\main.py", line 105, in infer
    slct_logi_feat, slct_dets_feat, slct_output_dets = self.det_process(
  File "D:\anaconda\envs\paddleocr\lib\site-packages\lineless_table_rec\lineless_table_process.py", line 26, in __call__
    dets, keep, logi, cr = ctdet_4ps_decode(
  File "D:\anaconda\envs\paddleocr\lib\site-packages\lineless_table_rec\lineless_table_process.py", line 107, in ctdet_4ps_decode
    scores, inds, clses, ys, xs = _topk(heat, K=K)
  File "D:\anaconda\envs\paddleocr\lib\site-packages\lineless_table_rec\lineless_table_process.py", line 206, in _topk
    topk_inds = _gather_feat(topk_inds.reshape(batch, -1, 1), topk_ind).reshape(
  File "D:\anaconda\envs\paddleocr\lib\site-packages\lineless_table_rec\lineless_table_process.py", line 251, in _gather_feat
    feat = _gather(feat, 1, ind)
  File "D:\anaconda\envs\paddleocr\lib\site-packages\lineless_table_rec\lineless_table_process.py", line 277, in _gather
    raise TypeError("The values of index must be integers")
TypeError: The values of index must be integers

Traceback (most recent call last):
  File "D:\工作\checkreport_table\testcode\lineless.py", line 13, in <module>
    print(table_str)
KeyboardInterrupt

Process finished with exit code -1073741510 (0xC000013A: interrupted by Ctrl+C)
```

请问是什么原因导致的呢

**Comment by nissansz** [comment]
解决了吗？也遇到了这个错误

\wired_table_rec>python test.py
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    table_str, elapse = table_rec(img_path)
  File "D:\Python3.8.8\lib\site-packages\wired_table_rec\main.py", line 45, in __call__
    polygons = self.table_line_rec(img)
  File "D:\Python3.8.8\lib\site-packages\wired_table_rec\table_line_rec.py", line 37, in __call__
    polygons = self.postprocess(pred)
  File "D:\Python3.8.8\lib\site-packages\wired_table_rec\table_line_rec.py", line 97, in postprocess
    bbox, _ = bbox_decode(hm[:, 0:1, :, :], c2v, reg=reg, K=self.K)
  File "D:\Python3.8.8\lib\site-packages\wired_table_rec\utils_table_line_rec.py", line 17, in bbox_decode
    scores, inds, clses, ys, xs = _topk(heat, K=K)
  File "D:\Python3.8.8\lib\site-packages\wired_table_rec\utils_table_line_rec.py", line 91, in _topk
    topk_inds = _gather_feat(topk_inds.reshape(batch, -1, 1), topk_ind).reshape(
  File "D:\Python3.8.8\lib\site-packages\wired_table_rec\utils_table_line_rec.py", line 134, in _gather_feat
    feat = _gather_np(feat, 1, ind)
  File "D:\Python3.8.8\lib\site-packages\wired_table_rec\utils_table_line_rec.py", line 160, in _gather_np
    raise TypeError("The values of index must be integers")
TypeError: The values of index must be integers

**Comment by SWHL** [comment]
请尝试用[tests目录](https://github.com/RapidAI/TableStructureRec/tree/main/tests/test_files/wired)下的表格图像试试，查看是否报同样的错误？

如果不报错误，说明是提供图像的问题，请方便的话，提供测试的图，便于我这里复现。

**Comment by nissansz** [comment]
所有图片都是这个报错

**Comment by nissansz** [comment]
系统是win10

**Comment by SWHL** [comment]
我这里测试的[这张图](https://github.com/RapidAI/TableStructureRec/blob/main/tests/test_files/wired/row_span.png)，并没有报这个错误。

测试代码：
```python
from pathlib import Path
from wired_table_rec import WiredTableRecognition


table_rec = WiredTableRecognition()

img_path = "tests/test_files/wired/row_span.png"
table_str, elapse = table_rec(img_path)
print(table_str)
print(elapse)

with open(f"{Path(img_path).stem}.html", "w", encoding="utf-8") as f:
    f.write(table_str)
```

输出结果：
```text
<html><body><table><tr><td rowspan="2">名称</td><td rowspan="2">产量（吨）</td><td colspan="2">环比</td></tr><tr><td>增长量（吨）</td><td>增长率（%)</td></tr><tr><td>荔枝</td><td>11</td><td></td><td>10</td></tr><tr><td>芒果</td><td></td><td></td><td>-10</td></tr><tr><td>香蕉</td><td></td><td></td><td>20</td></tr></table></body></html>
1.4739068750059232
```


如果想要我这里帮助排查问题，请给出详细运行代码、运行环境、python版本、所用图像。

**Comment by nissansz** [comment]
![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/f18c31f5-9255-49a2-8172-1975bfb05e56)
代码就是你上面的，图片也是你代码里的，就报错，python3.8.8

!!!
Traceback (most recent call last):
  File "F:/pycharm2020.2/RapidStructure-0.0.0/TableStructureRec-main/test.py", line 8, in <module>
    table_str, elapse = table_rec(img_path)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\main.py", line 45, in __call__
    polygons = self.table_line_rec(img)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\table_line_rec.py", line 37, in __call__
    polygons = self.postprocess(pred)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\table_line_rec.py", line 97, in postprocess
    bbox, _ = bbox_decode(hm[:, 0:1, :, :], c2v, reg=reg, K=self.K)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\utils_table_line_rec.py", line 17, in bbox_decode
    scores, inds, clses, ys, xs = _topk(heat, K=K)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\utils_table_line_rec.py", line 91, in _topk
    topk_inds = _gather_feat(topk_inds.reshape(batch, -1, 1), topk_ind).reshape(
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\utils_table_line_rec.py", line 134, in _gather_feat
    feat = _gather_np(feat, 1, ind)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\utils_table_line_rec.py", line 160, in _gather_np
    raise TypeError("The values of index must be integers")
TypeError: The values of index must be integers
!!!

**Comment by SWHL** [comment]
起码要让我这里能复现报错，才能知道是什么问题。

你这总是说报错报错，我也表示无能为力。

**Comment by nissansz** [comment]
代码就是你上面的，图片也是你代码里的，所有图片都是同样的报错，python3.8.8

TypeError: The values of index must be integers

**Comment by SWHL** [comment]
尝试python3.10 试试，如果也不行的话，那就不是我这里问题了。

**Comment by nissansz** [comment]
用colab运行没有报错

**Comment by nissansz** [comment]
![3](https://github.com/RapidAI/TableStructureRec/assets/41010669/a64bef44-87dc-4c43-a373-bd476d157b1e)
用倾斜表格测试，识别结果就不对了

![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/28ef9826-f770-4a4f-94f8-ecabece031ce)

**Comment by qqeip** [comment]
都报错 用样例的图片 3.7.6 3.10.10都一样
if index.dtype != np.dtype("int_"):
>           raise TypeError("The values of index must be integers")
E           TypeError: The values of index must be integers

E:\xx\site-packages\wired_table_rec-0.0.2-py3.10.egg\wired_table_rec\utils_table_line_rec.py:160: TypeError

**Reply by SWHL** [reply, marked as answer]
如果可以翻墙的话，建议去[Google Colab](https://colab.research.google.com/notebooks/welcome.ipynb#scrollTo=vR2GwbsQNnSS)上试试，上边有小伙伴已经反映Colab上没有问题了

如果不能翻墙，建议去百度的[AI Studio]()上试试，我在AI Studio上建了一个测试项目[表格识别测试](https://aistudio.baidu.com/projectdetail/7250181?contributionType=1&sUid=57084&shared=1&ts=1702376570638)

有些怀疑是Windows平台的原因。

如果你那里在linux上测试没有问题的话，方便给我提供一下，你出错的运行环境文件`requirements.txt`，我这里排查一下原因。

**Reply by nissansz** [reply]
有没有表格区域检测onnx模型？怎么使用？

**Comment by qqeip** [comment]
&nbsp;[RapidAI/RapidStructure] RapidLayout+RapidTable整合demo (PR #11)




---Original---
From: ***@***.***&gt;
Date: Tue, Dec 12, 2023 19:02 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [RapidAI/TableStructureRec] TypeError: The values of index mustbe integers (Discussion #2)




 
有没有表格区域检测onnx模型？怎么使用？

**Reply by nissansz** [reply]
准确率不太够

**Comment by qqeip** [comment]
Python 3.10.10 WIN10
[requirements.txt](https://github.com/RapidAI/TableStructureRec/files/13654378/requirements.txt)

**Reply by SWHL** [reply]
建议把numpy升级到最新试试
我手头没有win10的设备，暂时不能尝试。。。。

**Comment by qqeip** [comment]
aistudio上可以运行 但速度有点慢 要32秒
可能WIN10上不支持 我用pip install wired_table_rec也不行
<html><body><table><tr><td rowspan="2">名称</td><td rowspan="2">产量（吨）</td><td colspan="2">环比</td></tr><tr><td>增长量（吨）</td><td>增长率（%)</td></tr><tr><td>荔枝</td><td>11</td><td></td><td>10</td></tr><tr><td>芒果</td><td></td><td></td><td>-10</td></tr><tr><td>香蕉</td><td></td><td></td><td>20</td></tr></table></body></html>
32.92959946603514

**Reply by SWHL** [reply]
速度慢比较正常，AI Studio CPU差一些，后续有很多提速空间。比如用OpenVINO, onnxruntime-gpu版等等

**Comment by nissansz** [comment]
有没有完整前处理python代码？类似这些处理

https://medium.com/@Gekko_lab/make-your-ocr-results-more-accurate-part-ii-preprocessing-3d212ae16191

**Reply by qqeip** [reply]
cls det模型就是做这事的 详见paddleocr 或rapidocr

**Reply by nissansz** [reply]
好像不同。我想先前处理整页文档，不是检测文本行

**Comment by qqeip** [comment]
E:\OCR\TableStructureRec\wired_table_rec\utils_table_line_rec.py
def _gather_np(data, dim, index):里 把判断异常的两行注释掉就可以运行了
   #   if index.dtype != np.dtype("int_"):
    #     raise TypeError("The values of index must be integers")

python demo_wired.py
<html><body><table><tr><td rowspan="2">名称</td><td rowspan="2">产量（吨）</td><td colspan="2">环比</td></tr><tr><td>增长量（吨）</td><td>增长率（%)</td></tr><tr><td>荔枝</td><td>11</td><td></td><td>10</td></tr><tr><td>芒果</td><td></td><td></td><td>-10</td></tr><tr><td>香蕉</td><td></td><td></td><td>20</td></tr></table></body></html>
2.3827803004533052

**Reply by nissansz** [reply]
的确可以了。在哪里修改ocr引擎，有其他模型识别文字，放到表格里去？

**Reply by nissansz** [reply]
换了一个大表，又报其他错误了。
Traceback (most recent call last):
  File "F:/pycharm2020.2/RapidStructure-0.0.0/TableStructureRec-main/demo_wired.py", line 15, in <module>
    table_str, elapse = table_rec(img_path)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\main.py", line 45, in __call__
    polygons = self.table_line_rec(img)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\table_line_rec.py", line 43, in __call__
    polygons = merge_adjacent_polys(polygons)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\utils_table_recover.py", line 79, in merge_adjacent_polys
    polygons = combine_two_poly(polygons, idxs)
  File "F:\pycharm2020.2\RapidStructure-0.0.0\TableStructureRec-main\wired_table_rec\utils_table_recover.py", line 88, in combine_two_poly
    idxs = idxs.squeeze(0)
ValueError: cannot select an axis to squeeze out which has size not equal to one

**Reply by qqeip** [reply]
哪个图 我也只是看了样例 
main.py 里切换OCR
class WiredTableRecognition:
    def __init__(self, table_model_path: Union[str, Path] = default_model_path):
        self.load_img = LoadImage()
        self.table_line_rec = TableLineRecognition(str(table_model_path))
        self.table_recover = TableRecover()

        try:
            self.ocr = importlib.import_module("rapidocr_onnxruntime").RapidOCR()
        except ModuleNotFoundError:
            self.ocr = None

**Reply by SWHL** [reply]
@nissansz 建议问问题之前，自己多看看文档和源码。你这里问的很多问题，要不就是太大，要不就是一查文档就能解决

**Reply by nissansz** [reply]
你多试几张行数多或者复杂一点的图看看

**Reply by SWHL** [reply]
这个项目定位基于人家算法整合来的，并不保证解决所有问题。

遇到问题，自己多多尝试解决吧。可以一起商量，但我是为爱发电，没有一定帮你解决的义务哈。

希望你能明白这个事情。

**Reply by nissansz** [reply]
完全明白，只是反应一下实际结果和问题：这个模型有时甚至还没有之前的paddle表格模型准。许多也是识别不了的

**Reply by SWHL** [reply]
可以去看一下该算法对应的论文，其训练的数据集是自然场景的。测试时，可以根据手头表格场景，选用不同的表格识别模型。

Paddle的表格识别大部分都是基于印刷截屏的表格。如果你的表格是印刷截屏这种，那建议还是用Paddle的。

详细三种算法对比，参见[Docs](https://rapidai.github.io/TableStructureRec/docs/blog/table_rec_evaluate/)

**Reply by nissansz** [reply]
好的。 有没有什么更好的LayoutLM 的onnx模型可以用到radipocr?
https://zhuanlan.zhihu.com/p/486681158

**Reply by SWHL** [reply]
我这里没有整理过这块代码。欢迎贡献

**Reply by nissansz** [reply]
你训练过paddle表格模型吗？有训练出比paddle官网模型更高准确率的表格识别模型吗？
我尝试训练，都没官网准确率高

**Reply by SWHL** [reply]
@nissansz 
这个建议重新开一个Discussion，已经超出该Discussion的范围

**Reply by nissansz** [reply]
如何想可视化画出polygons，要怎么加代码？

main.py

        polygons = self.table_line_rec(img)
        if polygons is None:
            logging.warning("polygons is None.")
            return "", 0.0

        print(polygons)
        import cv2
        # 绘制四边形框
        cv2.polylines(img, polygons, isClosed=True, color=(0, 255, 0), thickness=2)

**Reply by nissansz** [reply]
写成这样，可以看到单元格了。但是有的png图片显示成黑底了
        print(polygons)
        import cv2
        import numpy as np
        # 重新整理polygons的形状
        polygons_int32 = polygons.astype(np.int32)
        # 绘制四边形框
        cv2.polylines(img, polygons_int32 , isClosed=True, color=(0, 255, 0), thickness=2)
        # 显示绘制结果
        cv2.imshow('Polygon', img)

明显的边框，有时识别不出来，导致识别结果错误
![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/00ab3f25-8825-47ed-bbb5-8a5cb0a2e01f)

**Reply by nissansz** [reply]
![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/9ce37e79-0ddf-44ba-b930-3918244f62e8)
多出很多空单元格

**Reply by nissansz** [reply]
倾斜表格的框是对的，但是整合成表格时出错了，不知道要怎么判断，更新，
![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/85619f5b-3a7a-4c7b-b579-651202eb47c8)
![image](https://github.com/RapidAI/TableStructureRec/assets/41010669/c6e92840-daf9-4db7-811e-17ef47d62294)
